[package] Add `deny` method to `PackageExporter`

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53233 [package] Add `deny` method to `PackageExporter`**
* #53232 [package] Add `allow_empty` flag to mock and extern

**Summary**
This commit adds a `deny` method to `PackageExporter` that allows
modules to be prohibited during the packaging process. A dependency on a
module matching the names or globs that `deny` was called with will
cause an exception to be raised.

**Test Plan**
This commit adds unit tests to `PackagingTest` for this new method:
`test_deny` and `test_deny_glob`.

**Fixes**
This commit fixes #53217.

Differential Revision: [D26834010](https://our.internmc.facebook.com/intern/diff/D26834010)

Commit:
[package] Add `deny` method to `PackageExporter`

**Summary**
This commit adds a `deny` method to `PackageExporter` that allows
modules to be prohibited during the packaging process. A dependency on a
module matching the names or globs that `deny` was called with will
cause an exception to be raised.

**Test Plan**
This commit adds unit tests to `PackagingTest` for this new method:
`test_deny` and `test_deny_glob`.

**Fixes**
This commit fixes #53217.

[ghstack-poisoned]

suo commented on torch/package/package_exporter.py:
```diff
@@ -419,6 +431,12 @@ def save_mock_module(self, module_name: str):
         is_package = hasattr(self._import_module(module_name), '__path__')
         self.save_source_string(module_name, _MOCK_IMPL, is_package, dependencies=False)
 
+    def reject_denied_module(self, module_name: str):
```
nit: prefix with `_` as this is a private method. The equivalents for mock and deny have a valid reason to be public (users may want to explicitly extern/mock a specific module) but that's not really the case here.

suo commented on torch/package/package_exporter.py:
```diff
@@ -419,6 +431,12 @@ def save_mock_module(self, module_name: str):
+        """Throw an exception containing a message that `module_name` was explicitly blocklisted via
+        `deny` and was still required during packaging.
+        """
+        raise RuntimeError(f"{module_name} was required during packaging but has been explicitly blocklisted")
```
similar to before, let's define a specific error class so it can be caught and handled separately by users, DeniedModuleError?

Commit:
Update on "[package] Add `deny` method to `PackageExporter`"


**Summary**
This commit adds a `deny` method to `PackageExporter` that allows
modules to be prohibited during the packaging process. A dependency on a
module matching the names or globs that `deny` was called with will
cause an exception to be raised.

**Test Plan**
This commit adds unit tests to `PackagingTest` for this new method:
`test_deny` and `test_deny_glob`.

**Fixes**
This commit fixes #53217.

[ghstack-poisoned]
